framework/include/utils/CircularBuffer.h

```diff
@@ -0,0 +1,155 @@
+ * Also means that operator[] works sequentially between 0 and size()
+ *
+ * It will also automatically grow larger if capacity is reached
+ */
+template <typename T>
+class CircularBuffer : public Buffer<T>
+{
+public:
+  /**
+   * Create an empty circular buffer
+   */
+  CircularBuffer();
+
+  /**
+   * Create a circular buffer with a specific capacity
+   */
+  CircularBuffer(const std::size_t capacity);
+
+  virtual void erase(const std::size_t num) override;
+  virtual void eraseChunk(const std::size_t chunk_size) override;
+
+  virtual typename Buffer<T>::iterator beginChunk(const std::size_t chunk_size) override;
+  virtual typename Buffer<T>::const_iterator
+  beginChunk(const std::size_t chunk_size) const override;
+  virtual typename Buffer<T>::iterator endChunk(const std::size_t chunk_size) override;
+  virtual typename Buffer<T>::const_iterator endChunk(const std::size_t chunk_size) const override;
+
+protected:
+  virtual std::size_t newEnd(const std::size_t new_end) override;
+};
+
+template <typename T>
+CircularBuffer<T>::CircularBuffer() : Buffer<T>()
+{
+}
+
+template <typename T>
+CircularBuffer<T>::CircularBuffer(const std::size_t capacity) : Buffer<T>(capacity)
+{
+}
+
+template <typename T>
+void
+CircularBuffer<T>::erase(const std::size_t num)
+{
+  mooseAssert(num <= this->size(), "Cannot erase past the last entry");
+
+  this->_begin_pos += num;
+
+  // If there's nothing in the buffer - let's reset the positions
+  if (this->_begin_pos == this->_end_pos)
+    this->clear();
+}
+
+template <typename T>
+void
+CircularBuffer<T>::eraseChunk(const std::size_t chunk_size)
+{
+  if (chunk_size > this->size())
+    this->erase(this->size());
+  else
+    this->erase(chunk_size);
+}
+
+template <typename T>
+typename Buffer<T>::iterator
+CircularBuffer<T>::beginChunk(const std::size_t /* chunk_size */)
+{
+  return this->begin();
+}
+
+template <typename T>
+typename Buffer<T>::const_iterator
+CircularBuffer<T>::beginChunk(const std::size_t /* chunk_size */) const
+{
+  return this->begin();
+}
+
+template <typename T>
+typename Buffer<T>::iterator
+CircularBuffer<T>::endChunk(const std::size_t chunk_size)
+{
+  if (chunk_size > this->size())
+    return this->end();
+  else
+    return this->begin() + chunk_size;
+}
+
+template <typename T>
+typename Buffer<T>::const_iterator
+CircularBuffer<T>::endChunk(const std::size_t chunk_size) const
+{
+  if (chunk_size > this->size())
+    return this->end();
+  else
+    return this->begin() + chunk_size;
+}
+
+template <typename T>
+std::size_t
+CircularBuffer<T>::newEnd(const std::size_t new_end)
+{
+  auto actual_new_end = new_end;
+
+  if (new_end > this->_data.size())
+  {
+    const auto new_size = new_end - this->_begin_pos;
+
+    // See if we need to grow our capacity
+    if (this->_begin_pos == 0) // If we're already using the beginning - just resize
+      this->_data.resize(std::max(new_size * 2, this->_data.size() * 2));
```

Comment:
this max is not necessary

begin_pos is 0, so new_size is new_end, and new_end is already bigger than data.size()